src/dachshund/candidate.rs

```diff
@@ -411,13 +465,39 @@ impl<'a, TGraph: GraphBase> Candidate<'a, TGraph> {
 
     /// counts the total number of ties between candidate's core nodes and non_cores
     pub fn count_ties_between_nodes(&self) -> CLQResult<usize> {
-        let mut num_ties: usize = 0;
-        for &non_core_id in &self.non_core_ids {
-            num_ties += self
-                .get_node(non_core_id)
-                .count_ties_with_ids(&self.core_ids);
+        Ok(self.ties_between_nodes)
+    }
+
+    // Update the count of ties between nodes to account for adding node_id. Can be called
+    // immediately before or immediately after inserting node into the set of ids.
+    fn increment_ties_between_nodes(&mut self, node_id: NodeId) {
+        let new_ties = if self.graph.get_node(node_id).is_core() {
+            self.get_node(node_id).count_ties_with_ids(&self.non_core_ids)
+        } else {
+            self.get_node(node_id).count_ties_with_ids(&self.core_ids)
+        };
+        self.ties_between_nodes += new_ties;
+    }
+
+    // Adjust the neighborhood property to account for adding added_node
+    fn adjust_neighborhood(&mut self, node_id: NodeId) {
+        let opposite_shore = if self.graph.get_node(node_id).is_core()
+            { &self.non_core_ids } else { &self.core_ids };
+
+        let neighbors : Vec<NodeId> = self.get_node(node_id)
+            .edges
+            .iter()
+            .map(|x| x.target_id)
+            .collect();
+
+        for target_id in neighbors {
+            if !opposite_shore.contains(&target_id) {
+                let counter = self.neighborhood.entry(target_id).or_insert(0);
+                *counter += if opposite_shore.contains(&target_id) {-1} else {1};
```

Comment:
isn't this if redundant with the if above? we shouldn't need to do anything if the opposite_shore already contains the target_id correct?

Reply:
You're right: the second if is redundant (and the -1 part is nonsense).